--- a/net/minecraft/server/network/ServerLessCommonPacketListenerImpl.java
+++ b/net/minecraft/server/network/ServerLessCommonPacketListenerImpl.java
@@ -27,8 +_,8 @@
     private static final int CLOSED_LISTENER_TIMEOUT = 15000;
     private static final Component TIMEOUT_DISCONNECTION_MESSAGE = Component.translatable("disconnect.timeout");
     static final Component DISCONNECT_UNEXPECTED_QUERY = Component.translatable("multiplayer.disconnect.unexpected_query_response");
-    protected final Connection connection;
-    private final boolean transferred;
+    public final Connection connection;
+    final boolean transferred;
     private long keepAliveTime;
     private boolean keepAlivePending;
     private long keepAliveChallenge;
