--- a/net/minecraft/world/inventory/DimensionControlMenu.java
+++ b/net/minecraft/world/inventory/DimensionControlMenu.java
@@ -78,6 +_,11 @@
     }
 
     @Override
+    public org.bukkit.inventory.InventoryView getBukkitView() {
+        return null; // TODO: implement later
+    }
+
+    @Override
     public ItemStack quickMoveStack(Player player, int index) {
         ItemStack itemStack = ItemStack.EMPTY;
         Slot slot = this.slots.get(index);
