--- a/net/minecraft/gametest/framework/GameTestServer.java
+++ b/net/minecraft/gametest/framework/GameTestServer.java
@@ -124,7 +_,7 @@
     }
 
     private GameTestServer(Thread thread, LevelStorageSource.LevelStorageAccess levelStorageAccess, Optional<String> optional, boolean flag) {
-        super(thread, levelStorageAccess, Proxy.NO_PROXY, DataFixers.getDataFixer(), NO_SERVICES);
+        super(null, null, thread, levelStorageAccess, Proxy.NO_PROXY, DataFixers.getDataFixer(), NO_SERVICES); // paper
         this.testSelection = optional;
         this.verify = flag;
     }
@@ -144,6 +_,15 @@
             chunkProgressListenerFactory,
             theGame1 -> new PlayerList(theGame1, this.playerDataStorage, 1) {}
         );
+        // Paper start
+        this.setPlayerList(new PlayerList(this, this.registries(), this.playerDataStorage, 1) {
+            @Override
+            public void loadAndSaveFiles() {
+                throw new UnsupportedOperationException("Should not be called in a GameTestServer");
+            }
+        });
+        this.loadLevel("blah");
+        // Paper end
         ServerLevel serverLevel = theGame.overworld();
         this.testBatches = this.evaluateTestsToRun(serverLevel);
         LOGGER.info("Started game test server");
@@ -327,6 +_,13 @@
     public boolean shouldInformAdmins() {
         return false;
     }
+
+    // Paper start
+    @Override
+    public org.bukkit.command.CommandSender getBukkitSender(final net.minecraft.commands.CommandSourceStack wrapper) {
+        throw new UnsupportedOperationException("Not supported.");
+    }
+    // Paper end
 
     @Override
     public boolean isSingleplayerOwner(GameProfile profile) {
