--- a/net/minecraft/world/item/FlintAndSteelItem.java
+++ b/net/minecraft/world/item/FlintAndSteelItem.java
@@ -25,10 +_,10 @@
 
     @Override
     public InteractionResult useOn(UseOnContext context) {
-        return (InteractionResult)(ignite(context) ? InteractionResult.SUCCESS : InteractionResult.FAIL);
+        return ignite(context);
     }
 
-    public static boolean ignite(UseOnContext useOnContext) {
+    public static InteractionResult ignite(UseOnContext useOnContext) {
         Player player = useOnContext.getPlayer();
         Level level = useOnContext.getLevel();
         BlockPos clickedPos = useOnContext.getClickedPos();
@@ -36,6 +_,12 @@
         if (!CampfireBlock.canLight(blockState) && !CandleBlock.canLight(blockState) && !CandleCakeBlock.canLight(blockState)) {
             BlockPos blockPos = clickedPos.relative(useOnContext.getClickedFace());
             if (BaseFireBlock.canBePlacedAt(level, blockPos, useOnContext.getHorizontalDirection())) {
+                // CraftBukkit start - Store the clicked block
+                if (org.bukkit.craftbukkit.event.CraftEventFactory.callBlockIgniteEvent(level, blockPos, org.bukkit.event.block.BlockIgniteEvent.IgniteCause.FLINT_AND_STEEL, player).isCancelled()) {
+                    useOnContext.getItemInHand().hurtAndBreak(1, player, LivingEntity.getSlotForHand(useOnContext.getHand()));
+                    return InteractionResult.PASS;
+                }
+                // CraftBukkit end
                 level.playSound(player, blockPos, SoundEvents.FLINTANDSTEEL_USE, SoundSource.BLOCKS, 1.0F, level.getRandom().nextFloat() * 0.4F + 0.8F);
                 BlockState state = BaseFireBlock.getState(level, blockPos);
                 level.setBlock(blockPos, state, 11);
@@ -46,11 +_,17 @@
                     itemInHand.hurtAndBreak(1, player, LivingEntity.getSlotForHand(useOnContext.getHand()));
                 }
 
-                return true;
+                return InteractionResult.SUCCESS;
             } else {
-                return false;
+                return InteractionResult.FAIL;
             }
         } else {
+            // CraftBukkit start - Store the clicked block
+            if (org.bukkit.craftbukkit.event.CraftEventFactory.callBlockIgniteEvent(level, clickedPos, org.bukkit.event.block.BlockIgniteEvent.IgniteCause.FLINT_AND_STEEL, player).isCancelled()) {
+                useOnContext.getItemInHand().hurtAndBreak(1, player, LivingEntity.getSlotForHand(useOnContext.getHand()));
+                return InteractionResult.PASS;
+            }
+            // CraftBukkit end
             level.playSound(player, clickedPos, SoundEvents.FLINTANDSTEEL_USE, SoundSource.BLOCKS, 1.0F, level.getRandom().nextFloat() * 0.4F + 0.8F);
             level.setBlock(clickedPos, blockState.setValue(BlockStateProperties.LIT, true), 11);
             level.gameEvent(player, GameEvent.BLOCK_CHANGE, clickedPos);
@@ -58,7 +_,7 @@
                 useOnContext.getItemInHand().hurtAndBreak(1, player, LivingEntity.getSlotForHand(useOnContext.getHand()));
             }
 
-            return true;
+            return InteractionResult.SUCCESS;
         }
     }
 }
