--- a/net/minecraft/server/TheGame.java
+++ b/net/minecraft/server/TheGame.java
@@ -110,7 +_,7 @@
     private TheGame.ReloadableResources resources;
     private final StructureTemplateManager structureTemplateManager;
     private final WorldData worldData;
-    private final PotionBrewing potionBrewing;
+    public PotionBrewing potionBrewing;
     private FuelValues fuelValues;
     private final ServerScoreboard scoreboard;
     @Nullable
@@ -132,7 +_,7 @@
         this.registries = worldStem.registries();
         this.worldData = worldStem.worldData();
         this.playerList = function.apply(this);
-        this.scoreboard = new ServerScoreboard(this.playerList::broadcastAll);
+        this.scoreboard = new ServerScoreboard(this); // Paper - use theGame directly
         if (!this.registries.compositeAccess().lookupOrThrow(Registries.LEVEL_STEM).containsKey(LevelStem.OVERWORLD)) {
             throw new IllegalStateException("Missing Overworld dimension data");
         } else {
